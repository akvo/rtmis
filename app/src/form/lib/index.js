import * as Yup from 'yup';
import { i18n } from '../../lib';

const intersection = (array1, array2) => {
  const set1 = new Set(array1);
  const result = [];
  for (const item of array2) {
    if (set1.has(item)) {
      result.push(item);
    }
  }
  return result;
};

const getDependencyAncestors = (questions, current, dependencies) => {
  const ids = dependencies.map((x) => x.id);
  const ancestors = questions.filter((q) => ids.includes(q.id)).filter((q) => q?.dependency);
  if (ancestors.length) {
    dependencies = ancestors.map((x) => x.dependency);
    current = [current, ...dependencies].flatMap((x) => x);
    ancestors.forEach((a) => {
      if (a?.dependency) {
        current = getDependencyAncestors(questions, current, a.dependency);
      }
    });
  }
  return current;
};

export const transformForm = (forms, lang = 'en') => {
  const nonEnglish = lang !== 'en';
  if (nonEnglish) {
    forms = i18n.transform(lang, forms);
  }
  const questions = forms.question_group
    .map((x) => {
      return x.question;
    })
    .flatMap((q) => q)
    .map((q) => (nonEnglish ? i18n.transform(lang, q) : q))
    .map((q) => {
      if (q.type === 'option' || q.type === 'multiple_option') {
        const options = q.option
          .map((o) => ({ ...o, label: o.name }))
          .map((o) => {
            return nonEnglish ? i18n.transform(lang, o) : o;
          });
        return {
          ...q,
          option: options.sort((a, b) => a.order - b.order),
        };
      }
      if (q?.tooltip) {
        const transTooltip = nonEnglish ? i18n.transform(lang, q.tooltip) : q.tooltip;
        return {
          ...q,
          tooltip: transTooltip,
        };
      }
      return q;
    });

  const transformed = questions.map((x) => {
    let requiredSignTemp = x?.requiredSign || null;
    if (x?.required && !x?.requiredSign) {
      requiredSignTemp = '*';
    }
    if (x?.dependency) {
      return {
        ...x,
        requiredSign: requiredSignTemp,
        dependency: getDependencyAncestors(questions, x.dependency, x.dependency),
      };
    }
    return {
      ...x,
      requiredSign: requiredSignTemp,
    };
  });

  return {
    ...forms,
    question_group: forms.question_group
      .sort((a, b) => a.order - b.order)
      .map((qg, qgi) => {
        let repeat = {};
        let repeats = {};
        if (qg?.repeatable) {
          repeat = { repeat: 1 };
          repeats = { repeats: [0] };
        }
        const translatedQg = nonEnglish ? i18n.transform(lang, qg) : qg;
        return {
          ...translatedQg,
          ...repeat,
          ...repeats,
          id: qg?.id || qgi,
          question: qg.question
            ?.sort((a, b) => a.order - b.order)
            ?.map((q) => {
              return transformed.find((t) => t.id === q.id);
            }),
        };
      }),
  };
};

export const modifyDependency = ({ question }, { dependency }, repeat) => {
  const questions = question.map((q) => q.id);
  return dependency.map((d) => {
    if (questions.includes(d.id) && repeat) {
      return { ...d, id: `${d.id}-${repeat}` };
    }
    return d;
  });
};

export const validateDependency = (dependency, value) => {
  if (dependency?.options && typeof value !== 'undefined') {
    if (typeof value === 'string') {
      value = [value];
    }
    return intersection(dependency.options, value)?.length > 0;
  }
  let valid = false;
  if (dependency?.min) {
    valid = value >= dependency.min;
  }
  if (dependency?.max) {
    valid = value <= dependency.max;
  }
  if (dependency?.equal) {
    valid = value === dependency.equal;
  }
  if (dependency?.notEqual) {
    valid = value !== dependency.notEqual && !!value;
  }
  return valid;
};

{
  /** TODO:: Delete if no longer required
export const generateValidationSchema = (forms) => {
  const questions = forms?.question_group
    ?.map((qg) => {
      const qs = qg.question.map((q) => {
        return {
          ...q,
          group: qg,
        };
      });
      return qs;
    })
    .flat();
  const schema = Yup.object().shape(
    questions
      .filter((q) => !q?.dependency)
      .reduce((res, curr) => {
        const { id, name, type, required, rule, group, dependency } = curr;
        const requiredError = `${name} is required.`;
        let yupType;
        switch (type) {
          case 'number':
            // number rules
            yupType = Yup.number();
            if (rule?.min) {
              yupType = yupType.min(rule.min);
            }
            if (rule?.max) {
              yupType = yupType.max(rule.max);
            }
            if (!rule?.allowDecimal) {
              // by default decimal is allowed
              yupType = yupType.integer();
            }
            break;
          case 'date':
            yupType = Yup.date();
            break;
          case 'option':
            yupType = Yup.array();
            break;
          case 'multiple_option':
            yupType = Yup.array();
            break;
          default:
            yupType = Yup.string();
            break;
        }
        // dependency & required check
        let yupRule = yupType;
        if (required && dependency && dependency?.length) {
          const repeat = 0;
          const modifiedDependency = modifyDependency(group, curr, repeat);
          modifiedDependency.forEach(({ id, options }) => {
            yupRule = yupRule.when(`${id}`, {
              is: (value) => intersection(value, options),
              then: yupType.required(requiredError),
            });
          });
        } else if (required) {
          yupRule = yupRule.required(requiredError);
        }
        return {
          ...res,
          [id]: yupRule,
        };
      }, {}),
  );
  return schema;
};
*/
}

export const generateValidationSchemaFieldLevel = (currentValue, field) => {
  const { name, type, required, rule } = field;
  let yupType;
  switch (type) {
    case 'number':
      // number rules
      const isEmpyCurrentValue = currentValue === '';
      yupType = isEmpyCurrentValue ? Yup.string() : Yup.number();
      if (!isEmpyCurrentValue && rule?.min) {
        yupType = yupType.min(rule.min);
      }
      if (!isEmpyCurrentValue && rule?.max) {
        yupType = yupType.max(rule.max);
      }
      if (!isEmpyCurrentValue && !rule?.allowDecimal) {
        // by default decimal is allowed
        yupType = yupType.integer();
      }
      break;
    case 'date':
      yupType = Yup.date();
      break;
    case 'option':
      yupType = Yup.array();
      break;
    case 'multiple_option':
      yupType = Yup.array();
      break;
    case 'cascade':
      yupType = Yup.array();
      break;
    case 'geo':
      yupType = Yup.array();
      break;
    default:
      yupType = Yup.string();
      break;
  }
  if (required) {
    const requiredError = `${name} is required.`;
    yupType = yupType.required(requiredError);
  }
  try {
    yupType.validateSync(currentValue);
  } catch (error) {
    return error.message;
  }
};

export const generateDataPointName = (forms, currentValues, cascades = {}) => {
  const dataPointNameValues = forms?.question_group?.length
    ? forms.question_group
        .filter((qg) => !qg?.repeatable)
        .flatMap((qg) => qg.question.filter((q) => q?.meta))
        ?.map((q) => {
          const defaultValue = currentValues?.[q.id] || null;
          const value = q.type === 'cascade' ? cascades?.[q.id] || defaultValue : defaultValue;
          return { id: q.id, type: q.type, value };
        })
    : [];

  const dpName = dataPointNameValues
    .filter((d) => d.type !== 'geo' && (d.value || d.value === 0))
    .map((x) => x.value)
    .join(' - ');
  const [lat, lng] = dataPointNameValues.find((d) => d.type === 'geo')?.value || [];
  const dpGeo = lat && lng ? `${lat}|${lng}` : null;
  return { dpName, dpGeo };
};

export const getCurrentTimestamp = () => Math.floor(Date.now() / 1000);

export const getDurationInMinutes = (startTime) => {
  // Get the current timestamp in seconds
  const endTime = getCurrentTimestamp();
  // Calculate the duration in seconds
  const durationInSeconds = endTime - startTime;

  return Math.floor(durationInSeconds / 60);
};
