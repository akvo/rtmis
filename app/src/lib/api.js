import axios from 'axios';

export const config = {
  baseURL: null,
  headers: {
    'Content-Type': 'application/json',
  },
};

const API = () => {
  const getConfig = () => {
    let current = config;
    if (api.baseURL) {
      current = {
        ...config,
        baseURL: api.baseURL,
      };
    }
    return api?.token
      ? {
          ...current,
          headers: { ...config.headers, Authorization: `Bearer ${api.token}` },
        }
      : current;
  };
  return {
    get: (url, config = {}) => axios({ url, ...getConfig(), ...config }),
    post: (url, data, config = {}) =>
      axios({
        url,
        method: 'POST',
        data,
        ...{
          ...getConfig(),
          headers: {
            ...getConfig().headers,
            ...config?.headers,
          },
        },
      }),
    put: (url, data, config) => axios({ url, method: 'PUT', data, ...getConfig(), ...config }),
    patch: (url, data, config) => axios({ url, method: 'PATCH', data, ...getConfig(), ...config }),
    delete: (url) => axios({ url, method: 'DELETE', ...getConfig() }),
    setToken: (token) => {
      api.token = token;
    },
    setServerURL: (serverURL) => {
      api.baseURL = serverURL;
    },
    getConfig,
  };
};

const api = API();

export default api;
