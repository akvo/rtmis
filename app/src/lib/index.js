export { default as api } from './api';
export { default as loc } from './loc';
export { default as notification } from './notification';
export { default as backgroundTask } from './background-task';
export { default as i18n } from './i18n';
export { default as cascades } from './cascades';
