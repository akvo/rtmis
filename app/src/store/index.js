export { default as AuthState } from './auth';
export { default as BuildParamsState } from './buildParams';
export { default as FormState } from './forms';
export { default as UIState } from './ui';
export { default as UserState } from './users';
