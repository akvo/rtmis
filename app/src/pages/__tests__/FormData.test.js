import React, { useState } from 'react';
import { render, waitFor, fireEvent, act, renderHook } from '@testing-library/react-native';
import axios from 'axios';

import FormDataPage from '../FormData';
import crudDataPoints from '../../database/crud/crud-datapoints';
import { useNavigation } from '@react-navigation/native';
import { backgroundTask } from '../../lib';
import { FormState, UIState } from '../../store';
import api from '../../lib/api';

jest.mock('@react-navigation/native');
jest.mock('../../database/crud/crud-datapoints');
jest.mock('../../lib/background-task.js');
jest.mock('../../lib/api');
jest.mock('axios', () => ({
  all: jest.fn(() => Promise.resolve([{ data: { file: '/images/photo_111_123_xyz.jpeg' } }])),
}));

describe('FormDataPage', () => {
  beforeAll(() => {
    FormState.update((s) => {
      s.form = {
        json: JSON.stringify({
          question_group: [
            {
              question: [],
            },
          ],
        }),
      };
    });
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('renders correctly', async () => {
    const mockData = [
      {
        id: 1,
        name: 'Datapoint 1',
        createdAt: '2023-07-18T12:34:56.789Z',
        duration: 145,
        syncedAt: '2023-07-18T13:00:00.000Z',
      },
    ];

    crudDataPoints.selectDataPointsByFormAndSubmitted.mockResolvedValue(mockData);
    const tree = render(<FormDataPage />);
    await waitFor(() => expect(tree.toJSON()).toMatchSnapshot());
  });

  it('should show list of submitted datapoints', async () => {
    const mockRoute = {
      params: {
        id: 123,
        name: 'Form Name',
        showSubmitted: true,
      },
    };

    const mockData = [
      {
        id: 1,
        name: 'Datapoint 1',
        createdAt: '2023-07-18T12:34:56.789Z',
        duration: 145,
        syncedAt: '2023-07-18T13:00:00.000Z',
        submitted: 1,
      },
    ];

    crudDataPoints.selectDataPointsByFormAndSubmitted.mockResolvedValue(mockData);

    const wrapper = render(<FormDataPage route={mockRoute} />);

    await waitFor(() => {
      expect(wrapper.getByText('Form Name')).toBeTruthy();
      const list0 = wrapper.getByTestId('card-touchable-0');
      expect(list0.props.children[0].props.title).toEqual('Datapoint 1');
      expect(list0.props.children[0].props.subTitles[0]).toEqual('Created: 18/07/2023 12:34 PM');
      expect(list0.props.children[0].props.subTitles[1]).toEqual('Survey duration: 02h 25m');
      expect(list0.props.children[0].props.subTitles[2]).toEqual('Synced: 18/07/2023 01:00 PM');
    });
  });

  it('should show list of saved datapoints', async () => {
    const mockRoute = {
      params: {
        id: 123,
        name: 'Form Name',
        showSubmitted: false,
      },
    };

    const mockData = [
      {
        id: 1,
        name: 'Datapoint 1',
        createdAt: '2023-07-18T12:34:56.789Z',
        duration: 145,
        syncedAt: null,
        submitted: 0,
      },
    ];

    crudDataPoints.selectDataPointsByFormAndSubmitted.mockResolvedValue(mockData);

    const wrapper = render(<FormDataPage route={mockRoute} />);

    await waitFor(() => {
      expect(wrapper.getByText('Form Name')).toBeTruthy();
      const list0 = wrapper.getByTestId('card-touchable-0');
      expect(list0.props.children[0].props.title).toEqual('Datapoint 1');
      expect(list0.props.children[0].props.subTitles[0]).toEqual('Created: 18/07/2023 12:34 PM');
      expect(list0.props.children[0].props.subTitles[1]).toEqual('Survey duration: 02h 25m');
      expect(list0.props.children[0].props.subTitles[2]).toEqual(undefined);
    });
  });

  it('should have search input field', () => {
    const mockData = [
      {
        id: 1,
        name: 'Datapoint 1',
        createdAt: '2023-07-18T12:34:56.789Z',
        duration: 145,
        syncedAt: '2023-07-18T13:00:00.000Z',
      },
    ];
    crudDataPoints.selectDataPointsByFormAndSubmitted.mockResolvedValue(mockData);
    const wrapper = render(<FormDataPage />);
    expect(wrapper.queryByTestId('search-bar')).toBeTruthy();
  });

  it('should filter list of datapoint by search value', async () => {
    const mockData = [
      {
        id: 1,
        name: 'Datapoint 1',
        createdAt: '2023-07-18T12:34:56.789Z',
        duration: 145,
        syncedAt: '2023-07-18T13:00:00.000Z',
      },
      {
        id: 2,
        name: 'Datapoint 2',
        createdAt: '2023-07-18T12:34:56.789Z',
        duration: 145,
        syncedAt: '2023-07-18T13:00:00.000Z',
      },
    ];

    crudDataPoints.selectDataPointsByFormAndSubmitted.mockResolvedValue(mockData);

    const wrapper = render(<FormDataPage />);

    const searchField = wrapper.getByTestId('search-bar');
    expect(searchField).toBeDefined();
    fireEvent.changeText(searchField, 'Datapoint 1');

    await waitFor(() => {
      const list0 = wrapper.queryByTestId('card-touchable-0');
      expect(list0).toBeTruthy();

      const list1 = wrapper.queryByTestId('card-touchable-1');
      expect(list1).toBeFalsy();
    });
  });

  it('should navigate to FormPage with correct route params when datapoint list pressed', async () => {
    const mockNavigation = useNavigation();
    const mockRoute = {
      params: {
        id: 123,
        name: 'Form Name',
        showSubmitted: false,
      },
    };

    const mockData = [
      {
        id: 1,
        name: 'Datapoint 1',
        createdAt: '2023-07-18T12:34:56.789Z',
        duration: 145,
        syncedAt: null,
        submitted: 0,
      },
    ];
    crudDataPoints.selectDataPointsByFormAndSubmitted.mockResolvedValue(mockData);

    const wrapper = render(<FormDataPage navigation={mockNavigation} route={mockRoute} />);

    await waitFor(() => {
      const cardElement = wrapper.getByTestId('card-touchable-0');
      fireEvent.press(cardElement);
    });

    expect(mockNavigation.navigate).toHaveBeenCalledWith('FormPage', {
      ...mockRoute.params,
      dataPointId: 1,
      newSubmission: false,
    });
  });

  it('should not render render sync button on Saved FormData page', async () => {
    const mockRoute = {
      params: {
        id: 123,
        name: 'Form Name',
        showSubmitted: false,
      },
    };

    const mockData = [
      {
        id: 1,
        name: 'Datapoint 1',
        createdAt: '2023-07-18T12:34:56.789Z',
        duration: 145,
        syncedAt: '2023-07-18T13:00:00.000Z',
        submitted: 1,
      },
    ];

    crudDataPoints.selectDataPointsByFormAndSubmitted.mockResolvedValue(mockData);

    const wrapper = render(<FormDataPage route={mockRoute} />);

    await waitFor(() => {
      expect(wrapper.getByText('Form Name')).toBeTruthy();
      // check sync button rendered
      expect(wrapper.queryByTestId('button-to-trigger-sync')).toBeFalsy();
      const list0 = wrapper.getByTestId('card-touchable-0');
      expect(list0).toBeTruthy();
    });
  });

  it('should render sync button on Submitted FormData page', async () => {
    const mockRoute = {
      params: {
        id: 123,
        name: 'Form Name',
        showSubmitted: true,
      },
    };

    const mockData = [
      {
        id: 1,
        name: 'Datapoint 1',
        createdAt: '2023-07-18T12:34:56.789Z',
        duration: 145,
        syncedAt: null,
        submitted: 1,
      },
    ];

    crudDataPoints.selectDataPointsByFormAndSubmitted.mockResolvedValue(mockData);

    const wrapper = render(<FormDataPage route={mockRoute} />);

    await waitFor(() => {
      expect(wrapper.getByText('Form Name')).toBeTruthy();
      // check sync button rendered
      const syncButtonElement = wrapper.getByTestId('button-to-trigger-sync');
      expect(syncButtonElement).toBeTruthy();
      expect(syncButtonElement.props.accessibilityState.disabled).toEqual(false);
      const list0 = wrapper.getByTestId('card-touchable-0');
      expect(list0).toBeTruthy();
    });
  });

  it('should disable sync button if no data on Submitted FormData page', async () => {
    const mockRoute = {
      params: {
        id: 123,
        name: 'Form Name',
        showSubmitted: true,
      },
    };

    const mockData = [
      {
        id: 1,
        name: 'Datapoint 1',
        createdAt: '2023-07-18T12:34:56.789Z',
        duration: 145,
        syncedAt: '2023-07-18T13:00:00.000Z',
        submitted: 1,
      },
    ];

    crudDataPoints.selectDataPointsByFormAndSubmitted.mockResolvedValue(mockData);

    const wrapper = render(<FormDataPage route={mockRoute} />);

    await waitFor(() => {
      expect(wrapper.getByText('Form Name')).toBeTruthy();
      // check sync button rendered
      const syncButtonElement = wrapper.getByTestId('button-to-trigger-sync');
      expect(syncButtonElement).toBeTruthy();
      expect(syncButtonElement.props.accessibilityState.disabled).toEqual(true);
      const list0 = wrapper.queryByTestId('card-touchable-0');
      expect(list0).toBeFalsy();
    });
  });

  it('should show sync confirmation dialog when sync button pressed and hide if confirmation denied', async () => {
    const mockRoute = {
      params: {
        id: 123,
        name: 'Form Name',
        showSubmitted: true,
      },
    };

    const mockData = [
      {
        id: 1,
        name: 'Datapoint 1',
        createdAt: '2023-07-18T12:34:56.789Z',
        duration: 145,
        syncedAt: null,
        submitted: 1,
      },
    ];

    crudDataPoints.selectDataPointsByFormAndSubmitted.mockResolvedValue(mockData);

    const wrapper = render(<FormDataPage route={mockRoute} />);

    await waitFor(() => expect(wrapper.getByText('Form Name')).toBeTruthy());

    // check sync button rendered
    const syncButtonElement = wrapper.getByTestId('button-to-trigger-sync');
    expect(syncButtonElement).toBeTruthy();
    fireEvent.press(syncButtonElement);

    const dialogElement = wrapper.queryByTestId('sync-confirmation-dialog');
    expect(dialogElement).toBeTruthy();

    await waitFor(() => expect(dialogElement.props.visible).toEqual(true));

    const textConfirmationElement = wrapper.queryByTestId('sync-confirmation-text');
    expect(textConfirmationElement).toBeTruthy();
    const okButtonElement = wrapper.queryByTestId('sync-confirmation-ok');
    expect(okButtonElement).toBeTruthy();
    const cancelButtonElement = wrapper.queryByTestId('sync-confirmation-cancel');
    expect(cancelButtonElement).toBeTruthy();

    fireEvent.press(cancelButtonElement);

    await waitFor(() => expect(dialogElement.props.visible).toEqual(false));
  });

  it('should handle sync submission when sync confirmation granted', async () => {
    const mockRoute = {
      params: {
        id: 123,
        name: 'Form Name',
        showSubmitted: true,
      },
    };

    const mockData = [
      {
        id: 1,
        name: 'Datapoint 1',
        createdAt: '2023-07-18T12:34:56.789Z',
        duration: 145,
        syncedAt: null,
        submitted: 1,
      },
    ];

    crudDataPoints.selectDataPointsByFormAndSubmitted.mockResolvedValue(mockData);
    backgroundTask.syncFormSubmission.mockResolvedValue(() => Promise.resolve(true));

    const wrapper = render(<FormDataPage route={mockRoute} />);

    await waitFor(() => expect(wrapper.getByText('Form Name')).toBeTruthy());

    // check sync button rendered
    const syncButtonElement = wrapper.getByTestId('button-to-trigger-sync');
    expect(syncButtonElement).toBeTruthy();
    fireEvent.press(syncButtonElement);

    const dialogElement = wrapper.queryByTestId('sync-confirmation-dialog');
    expect(dialogElement).toBeTruthy();

    await waitFor(() => expect(dialogElement.props.visible).toEqual(true));

    const okButtonElement = wrapper.queryByTestId('sync-confirmation-ok');
    expect(okButtonElement).toBeTruthy();
    fireEvent.press(okButtonElement);

    await waitFor(() => {
      const dataPointListElement = wrapper.queryByTestId('data-point-list');

      expect(dataPointListElement).toBeFalsy();
      expect(backgroundTask.syncFormSubmission).toHaveBeenCalledTimes(1);
      expect(crudDataPoints.selectDataPointsByFormAndSubmitted).toHaveBeenCalledTimes(2);
    });
  });

  it('should go to ManageForm page when arrow back clicked', () => {
    const mockNavigation = useNavigation();
    const mockRoute = {
      params: {
        id: 123,
        name: 'Form Name',
        showSubmitted: false,
      },
    };

    const { getByTestId } = render(<FormDataPage navigation={mockNavigation} route={mockRoute} />);
    const arrowBackEl = getByTestId('arrow-back-button');
    expect(arrowBackEl).toBeDefined();
    fireEvent.press(arrowBackEl);

    expect(mockNavigation.navigate).toHaveBeenCalledWith('ManageForm', mockRoute.params);
  });

  it('should set currentValues & go to FormDataDetails when showSubmitted is true', async () => {
    const mockData = [
      {
        id: 1,
        name: 'Datapoint 1',
        createdAt: '2023-07-18T12:34:56.789Z',
        duration: 145,
        syncedAt: '2023-07-18T13:00:00.000Z',
        submitted: 1,
        json: '{"1": "John Doe"}',
      },
    ];

    crudDataPoints.selectDataPointsByFormAndSubmitted.mockResolvedValue(mockData);

    const mockNavigation = useNavigation();
    const mockRoute = {
      params: {
        id: 123,
        name: 'Form Name',
        showSubmitted: true,
      },
    };
    const { getByTestId } = render(<FormDataPage navigation={mockNavigation} route={mockRoute} />);

    await waitFor(() => {
      const list0 = getByTestId('card-touchable-0');
      expect(list0).toBeDefined();
      fireEvent.press(list0);
    });

    act(() => {
      const { json: valuesJSON } = mockData[0];
      FormState.update((s) => {
        const valuesParsed = JSON.parse(valuesJSON);
        s.currentValues =
          typeof valuesParsed === 'string' ? JSON.parse(valuesParsed) : valuesParsed;
      });
    });

    await waitFor(() => {
      expect(mockNavigation.navigate).toHaveBeenCalledWith('FormDataDetails', {
        name: 'Datapoint 1',
      });
    });
  });

  it('should be set isManualSynced true and syncedAt when the button sync clicked', async () => {
    const mockData = [
      {
        id: 1,
        name: 'Datapoint 1',
        createdAt: '2023-07-18T12:34:56.789Z',
        duration: 10,
        syncedAt: null,
        submitted: 1,
        json: '{"1": "John Doe"}',
      },
    ];

    crudDataPoints.selectDataPointsByFormAndSubmitted.mockResolvedValue(mockData);

    const mockNavigation = useNavigation();
    const mockRoute = {
      params: {
        id: 123,
        name: 'Form Name',
        showSubmitted: true,
      },
    };
    const { getByTestId, queryByTestId, getByText, rerender, debug, queryByText } = render(
      <FormDataPage navigation={mockNavigation} route={mockRoute} />,
    );

    await waitFor(() => expect(getByText('Form Name')).toBeTruthy());

    const syncButtonEl = getByTestId('button-to-trigger-sync');
    expect(syncButtonEl).toBeTruthy();
    expect(syncButtonEl.props.accessibilityState.disabled).toBeFalsy();
    fireEvent.press(syncButtonEl);

    const syncedAt = '2023-07-18T12:40:00.789Z';
    crudDataPoints.selectDataPointsByFormAndSubmitted.mockResolvedValue([
      { ...mockData[0], syncedAt },
    ]);

    const dialogElement = queryByTestId('sync-confirmation-dialog');
    expect(dialogElement).toBeTruthy();

    await waitFor(() => expect(dialogElement.props.visible).toEqual(true));

    const okButtonElement = queryByTestId('sync-confirmation-ok');
    expect(okButtonElement).toBeTruthy();
    fireEvent.press(okButtonElement);

    act(() => {
      UIState.update((s) => {
        s.isManualSynced = true;
      });
    });

    await crudDataPoints.selectDataPointsByFormAndSubmitted();

    rerender(<FormDataPage navigation={mockNavigation} route={mockRoute} />);

    await waitFor(() => {
      const { result } = renderHook(() => UIState.useState((s) => s.isManualSynced));
      expect(result.current).toBeTruthy();
      expect(queryByText('Synced: 18/07/2023')).toBeDefined();
    });
  });

  it('should upload photos first before syncing when some data point has them', async () => {
    api.post.mockImplementation(() =>
      Promise.resolve({ data: { file: '/images/photo-profile-xyz.jpeg' } }),
    );
    const mockedForm = {
      id: 1,
      name: 'Dummy Form',
      question_group: [
        {
          id: 11,
          name: 'General Info',
          question: [
            {
              id: 111,
              name: 'Photo profile',
              type: 'photo',
            },
          ],
        },
      ],
    };

    act(() => {
      const jsonString = JSON.stringify(mockedForm);
      FormState.update((s) => {
        s.form = {
          id: mockedForm.id,
          name: mockedForm.name,
          json: jsonString,
        };
      });
    });
    const mockData = [
      {
        id: 123,
        name: 'Datapoint with photo',
        createdAt: '2023-07-18T12:34:56.789Z',
        duration: 105,
        syncedAt: null,
        submitted: 1,
        json: '{"111": "file://photo-profile.jpeg"}',
      },
    ];

    crudDataPoints.selectDataPointsByFormAndSubmitted.mockResolvedValue(mockData);
    crudDataPoints.selectSubmissionToSync.mockResolvedValue(mockData);

    const mockNavigation = useNavigation();
    const mockRoute = {
      params: {
        id: mockedForm.id,
        name: mockedForm.name,
        showSubmitted: true,
      },
    };

    const { getByTestId, getByText, queryByText, queryByTestId } = render(
      <FormDataPage route={mockRoute} navigation={mockNavigation} />,
    );

    await waitFor(() => expect(getByText('Datapoint with photo')).toBeTruthy());

    const syncButtonElement = getByTestId('button-to-trigger-sync');
    expect(syncButtonElement.props.accessibilityState.disabled).toBeFalsy();
    fireEvent.press(syncButtonElement);

    const dialogElement = queryByTestId('sync-confirmation-dialog');
    expect(dialogElement).toBeTruthy();

    await waitFor(() => expect(dialogElement.props.visible).toEqual(true));

    const textConfirmationElement = queryByTestId('sync-confirmation-text');
    expect(textConfirmationElement).toBeTruthy();
    const okButtonElement = queryByTestId('sync-confirmation-ok');
    expect(okButtonElement).toBeTruthy();
    fireEvent.press(okButtonElement);

    await waitFor(() => {
      const titleDatapoint = queryByText('Datapoint with photo');
      expect(titleDatapoint).toBeDefined();

      expect(api.post).toHaveBeenCalledTimes(1);

      const mockedformData = new FormData();
      mockedformData.append('file', {
        uri: 'file://photo-profile.jpeg',
        name: 'photo_111_123.jpeg',
        type: 'image/jpeg',
      });
      expect(api.post).toHaveBeenCalledWith('/images', mockedformData, {
        headers: {
          Accept: 'application/json',
          'Content-Type': 'multipart/form-data',
        },
      });
    });
  });

  it('should not call handleOnSync when uploading photos failed', async () => {
    axios.all.mockImplementation(() => Promise.reject('Error'));
    api.post.mockImplementation(() =>
      Promise.resolve({ data: { file: '/images/photo-profile-xyz.jpeg' } }),
    );
    const mockedForm = {
      id: 1,
      name: 'Dummy Form',
      question_group: [
        {
          id: 11,
          name: 'General Info',
          question: [
            {
              id: 111,
              name: 'Photo profile',
              type: 'photo',
            },
          ],
        },
      ],
    };

    act(() => {
      const jsonString = JSON.stringify(mockedForm);
      FormState.update((s) => {
        s.form = {
          id: mockedForm.id,
          name: mockedForm.name,
          json: jsonString,
        };
      });
    });
    const mockData = [
      {
        id: 123,
        name: 'Datapoint with photo',
        createdAt: '2023-07-18T12:34:56.789Z',
        duration: 105,
        syncedAt: null,
        submitted: 1,
        json: '{"111": "file://photo-profile.jpeg"}',
      },
    ];

    crudDataPoints.selectDataPointsByFormAndSubmitted.mockResolvedValue(mockData);
    crudDataPoints.selectSubmissionToSync.mockResolvedValue(mockData);

    const mockNavigation = useNavigation();
    const mockRoute = {
      params: {
        id: mockedForm.id,
        name: mockedForm.name,
        showSubmitted: true,
      },
    };

    const { getByTestId, getByText, queryByText, queryByTestId } = render(
      <FormDataPage route={mockRoute} navigation={mockNavigation} />,
    );

    await waitFor(() => expect(getByText('Datapoint with photo')).toBeTruthy());

    const syncButtonElement = getByTestId('button-to-trigger-sync');
    fireEvent.press(syncButtonElement);

    const dialogElement = queryByTestId('sync-confirmation-dialog');

    await waitFor(() => expect(dialogElement.props.visible).toEqual(true));

    const okButtonElement = queryByTestId('sync-confirmation-ok');
    expect(okButtonElement).toBeTruthy();
    fireEvent.press(okButtonElement);

    const mockedHandleOnSync = jest.fn();
    await waitFor(() => {
      expect(mockedHandleOnSync).not.toHaveBeenCalled();
    });
  });
});
