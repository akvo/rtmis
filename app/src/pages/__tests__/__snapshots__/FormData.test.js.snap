// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormDataPage renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#f9fafb",
      "flex": 1,
      "flexDirection": "column",
      "flexWrap": "wrap",
    }
  }
  testID="stack-container"
>
  <View
    style={
      {
        "backgroundColor": "#393e42",
        "borderBottomColor": "#000",
        "borderBottomWidth": 1,
        "borderTopColor": "#000",
        "borderTopWidth": 1,
        "padding": 8,
      }
    }
    testID="RNE__SearchBar-wrapper"
  >
    <View
      style={
        {
          "paddingHorizontal": 0,
          "width": "100%",
        }
      }
      testID="RNE__Input__view-wrapper"
    >
      <View
        collapsable={false}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#303337",
            "borderBottomWidth": 0,
            "borderColor": "#86939e",
            "borderRadius": 3,
            "flexDirection": "row",
            "minHeight": 30,
            "overflow": "hidden",
            "transform": [
              {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginLeft": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              {
                "overflow": "hidden",
              }
            }
            testID="RNE__ICON__CONTAINER"
          >
            <View
              accessibilityRole="button"
              android_ripple={
                {
                  "borderless": false,
                  "color": "rgba(0, 0, 0, 0.3)",
                  "radius": -5,
                }
              }
              disabled={false}
              testID="RNE__ICON__CONTAINER_ACTION"
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
                testID="RNE__ICON"
              >
                <Text />
              </View>
            </View>
          </View>
        </View>
        <TextInput
          editable={true}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Search datapoint"
          placeholderTextColor="#86939e"
          platform="default"
          style={
            {
              "color": "#86939e",
              "flex": 1,
              "fontSize": 18,
              "marginLeft": 10,
              "minHeight": 40,
              "width": "100%",
            }
          }
          testID="search-bar"
          underlineColorAndroid="transparent"
          value={null}
        />
        <View
          style={
            {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginRight": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              {
                "flexDirection": "row",
              }
            }
          />
        </View>
      </View>
      <Text
        style={
          {
            "color": "#ff190c",
            "fontSize": 12,
            "height": 0,
            "margin": 0,
            "padding": 0,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      {
        "backgroundColor": "#f9fafb",
        "flex": 1,
        "flexDirection": "row",
        "flexWrap": "wrap",
      }
    }
    testID="stack-container"
  />
  <Modal
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    style={
      {
        "width": "100%",
      }
    }
    testID="sync-confirmation-dialog"
    transparent={true}
    visible={false}
  />
</View>
`;
